src/nix/flake.cc

```diff
@@ -333,11 +333,15 @@ struct CmdFlakeCheck : FlakeCommand
         auto checkOverlay = [&](const std::string & attrPath, Value & v, const Pos & pos) {
             try {
                 state->forceValue(v, pos);
-                if (!v.isLambda() || v.lambda.fun->matchAttrs || std::string(v.lambda.fun->arg) != "final")
-                    throw Error("overlay does not take an argument named 'final'");
+                if (!v.isLambda())
+                   throw Error("Expected a lambda for the overlay attribute, but got something else");
+                if (v.lambda.fun->matchAttrs)
+                    throw Error("By convention the attribute 'overlay' is expected to be a simple lambda starting with 'final: ', not an attribute set matching lambda");
```

Comment:
```suggestion
                    throw Error("the attribute 'overlay' must a function that takes a single argument named 'final ', not an attribute set");
```

It's not a convention but a requirement. Also instead of "lambda" it's better to say "function".

Reply:
What is responsible for this "requirement"?

Assuming we do need this, why say "function", a general term, when the value must be provided by a lambda? Or does Nix have another implementation of "function" that is acceptable here, but is not a lambda?

Reply:
This exact check is the requirement.